crates/core/c8y_api/src/http_proxy.rs

```diff
@@ -128,22 +129,29 @@ impl C8yMqttJwtTokenRetriever {
         // Ignore errors on this connection
         mqtt_con.errors.close();
 
-        mqtt_con
-            .published
-            .publish(mqtt_channel::Message::new(
-                &Topic::new_unchecked("c8y/s/uat"),
-                "".to_string(),
-            ))
-            .await?;
-
-        let token_smartrest =
-            match tokio::time::timeout(Duration::from_secs(10), mqtt_con.received.next()).await {
-                Ok(Some(msg)) => msg.payload_str()?.to_string(),
+        tokio::time::sleep(Duration::from_millis(20)).await;
+        for _ in 0..3 {
+            mqtt_con
+                .published
+                .publish(mqtt_channel::Message::new(
+                    &Topic::new_unchecked("c8y/s/uat"),
+                    "".to_string(),
+                ))
+                .await?;
+            debug!("JWT token requested");
+
+            match tokio::time::timeout(Duration::from_secs(20), mqtt_con.received.next()).await {
+                Ok(Some(msg)) => {
+                    debug!("JWT token received");
+                    let token_smartrest = msg.payload_str()?.to_string();
+                    return Ok(SmartRestJwtResponse::try_new(&token_smartrest)?);
+                }
                 Ok(None) => return Err(JwtError::NoJwtReceived),
-                Err(_elapsed) => return Err(JwtError::NoJwtReceived),
+                Err(_elapsed) => continue,
             };
+        }
 
-        Ok(SmartRestJwtResponse::try_new(&token_smartrest)?)
+        Err(JwtError::NoJwtReceived)
```

Comment:
How about logging the errors from the `mqtt_con.errors` channel as well before we return so that we know if there were some other connection errors as well?

Reply:
Fixed with https://github.com/thin-edge/thin-edge.io/pull/1977/commits/4e0c849956a1cb6f9a42897e1fe579fbad6ac84c